src/core/datatypes/trees/Clade.h

```diff
@@ -87,21 +88,24 @@ namespace RevBayesCore {
     private: 
         
         // members
-        double                                      age;
+        double                                      age = 0.0;
         RbBitSet                                    bitset;
         std::string                                 clade_name;
-        int                                         num_missing;
+        int                                         num_missing = 0;
         std::set<Taxon>                             mrca;
         std::vector<Taxon>                          taxa;
-        bool                                        is_negative_constraint;
-        bool                                        is_optional_match;
-        std::vector<Clade>                          optional_constraints;
-        
+        bool                                        is_negative_constraint = false;
+        boost::optional<std::vector<Clade>>         optional_constraints;
```

Comment:
what is this boost optional? Shouldn't a plain pointer work? Just thinking that not introducing more special language features makes the code harder to read. Just my old school way ...

Reply:
Hmm... so the optional<T> is definitely an improvement over having a separate boolean flag, because (i) we don't have to worry about keeping the flag in sync and (ii) we don't allocate a clade if we aren't using it.  I used boost::optional here, because std::optional<T> isn't available until c++17. 

std::optional<T> is different than pointers because it allocates on the stack, instead of on the heap.  It replaces the common practice of i.e. returning something like `-1` to indicate "no value", which can lead to mistakes.  I think it makes code _easier_ to read (after you get used to it) by explicitly stating that we may-or-may-not have a value.  For example, you can write `if (optional_constraints.has_value())`...

I would argue that std::optional<T> is one of the few features that is definitely worth the time to learn (See https://en.cppreference.com/w/cpp/utility/optional).

I also found this: https://devblogs.microsoft.com/cppblog/stdoptional-how-when-and-why/